File: internal/module/registry.go
package module

import (
	"fmt"
	"sort"
	"sync"
)

// ModuleInfo holds metadata about a registered module, suitable for
// display in CLI listings and API responses.
type ModuleInfo struct {
	ID                   string
	Name                 string
	Version              string
	Type                 string // "collector" or "tester"
	SourceSystem         string
	SafetyClassification string // empty for collectors, classification for testers
	EnvironmentScope     string // empty for collectors, scope for testers
}

// Registry holds all registered modules, providing thread-safe access
// to collectors and testers by their unique identifiers.
type Registry struct {
	mu         sync.RWMutex
	collectors map[string]Collector
	testers    map[string]Tester
}

// NewRegistry creates an empty module registry.
func NewRegistry() *Registry {
	return &Registry{
		collectors: make(map[string]Collector),
		testers:    make(map[string]Tester),
	}
}

// RegisterCollector adds a collector to the registry.
func (r *Registry) RegisterCollector(c Collector) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.collectors[c.ID()] = c
}

// RegisterTester adds a tester to the registry.
func (r *Registry) RegisterTester(t Tester) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.testers[t.ID()] = t
}

// GetCollector returns the collector with the given ID or an error if not found.
func (r *Registry) GetCollector(id string) (Collector, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	c, ok := r.collectors[id]
	if !ok {
		return nil, fmt.Errorf("collector %q not found", id)
	}
	return c, nil
}

// GetTester returns the tester with the given ID or an error if not found.
func (r *Registry) GetTester(id string) (Tester, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	t, ok := r.testers[id]
	if !ok {
		return nil, fmt.Errorf("tester %q not found", id)
	}
	return t, nil
}

// ListCollectors returns all registered collectors.
func (r *Registry) ListCollectors() []Collector {
	r.mu.RLock()
	defer r.mu.RUnlock()
	result := make([]Collector, 0, len(r.collectors))
	for _, c := range r.collectors {
		result = append(result, c)
	}
	return result
}

// ListTesters returns all registered testers.
func (r *Registry) ListTesters() []Tester {
	r.mu.RLock()
	defer r.mu.RUnlock()
	result := make([]Tester, 0, len(r.testers))
	for _, t := range r.testers {
		result = append(result, t)
	}
	return result
}

// ListAll returns all registered modules (both collectors and testers).
func (r *Registry) ListAll() []Module {
	r.mu.RLock()
	defer r.mu.RUnlock()
	result := make([]Module, 0, len(r.collectors)+len(r.testers))
	for _, c := range r.collectors {
		result = append(result, c)
	}
	for _, t := range r.testers {
		result = append(result, t)
	}
	return result
}

// GetModule returns a Module (collector or tester) by ID, or an error if not found.
func (r *Registry) GetModule(id string) (Module, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	if c, ok := r.collectors[id]; ok {
		return c, nil
	}
	if t, ok := r.testers[id]; ok {
		return t, nil
	}
	return nil, fmt.Errorf("module %q not found", id)
}

// ListModules returns metadata for all registered modules, sorted by ID.
func (r *Registry) ListModules() []ModuleInfo {
	r.mu.RLock()
	defer r.mu.RUnlock()

	infos := make([]ModuleInfo, 0, len(r.collectors)+len(r.testers))

	for _, c := range r.collectors {
		infos = append(infos, ModuleInfo{
			ID:           c.ID(),
			Name:         c.Name(),
			Version:      c.Version(),
			Type:         "collector",
			SourceSystem: c.SourceSystem(),
		})
	}

	for _, t := range r.testers {
		infos = append(infos, ModuleInfo{
			ID:                   t.ID(),
			Name:                 t.Name(),
			Version:              t.Version(),
			Type:                 "tester",
			SourceSystem:         t.SourceSystem(),
			SafetyClassification: string(t.SafetyClass()),
			EnvironmentScope:     string(t.EnvironmentScope()),
		})
	}

	sort.Slice(infos, func(i, j int) bool {
		return infos[i].ID < infos[j].ID
	})

	return infos
}

// ListByType returns metadata for modules of the given type ("collector" or "tester"),
// sorted by ID.
func (r *Registry) ListByType(moduleType string) []ModuleInfo {
	all := r.ListModules()
	filtered := make([]ModuleInfo, 0)
	for _, info := range all {
		if info.Type == moduleType {
			filtered = append(filtered, info)
		}
	}
	return filtered
}

// ListBySourceSystem returns metadata for modules targeting the given source system,
// sorted by ID.
func (r *Registry) ListBySourceSystem(system string) []ModuleInfo {
	all := r.ListModules()
	filtered := make([]ModuleInfo, 0)
	for _, info := range all {
		if info.SourceSystem == system {
			filtered = append(filtered, info)
		}
	}
	return filtered
}

File: src/module/registry.rs
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

use anyhow::{anyhow, Result};

use super::{Module, Observer, Tester};

/// Metadata about a registered module, suitable for CLI listings and API responses.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ModuleInfo {
    pub id: String,
    pub name: String,
    pub version: String,
    /// "observer" or "tester"
    pub module_type: String,
    pub source_system: String,
    /// Empty for observers.
    pub safety_classification: String,
    /// Empty for observers.
    pub environment_scope: String,
}

/// Thread-safe registry of all registered observers and testers.
#[derive(Default)]
pub struct Registry {
    observers: RwLock<HashMap<String, Arc<dyn Observer>>>,
    testers: RwLock<HashMap<String, Arc<dyn Tester>>>,
}

impl Registry {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn register_observer(&self, c: Arc<dyn Observer>) {
        self.observers
            .write()
            .unwrap()
            .insert(c.id().to_string(), c);
    }

    pub fn register_tester(&self, t: Arc<dyn Tester>) {
        self.testers.write().unwrap().insert(t.id().to_string(), t);
    }

    pub fn get_observer(&self, id: &str) -> Result<Arc<dyn Observer>> {
        self.observers
            .read()
            .unwrap()
            .get(id)
            .cloned()
            .ok_or_else(|| anyhow!("observer {id:?} not found"))
    }

    pub fn get_tester(&self, id: &str) -> Result<Arc<dyn Tester>> {
        self.testers
            .read()
            .unwrap()
            .get(id)
            .cloned()
            .ok_or_else(|| anyhow!("tester {id:?} not found"))
    }

    pub fn get_module(&self, id: &str) -> Result<Arc<dyn Module>> {
        if let Ok(c) = self.get_observer(id) {
            return Ok(c as Arc<dyn Module>);
        }
        if let Ok(t) = self.get_tester(id) {
            return Ok(t as Arc<dyn Module>);
        }
        Err(anyhow!("module {id:?} not found"))
    }

    pub fn list_observers(&self) -> Vec<Arc<dyn Observer>> {
        self.observers.read().unwrap().values().cloned().collect()
    }

    pub fn list_testers(&self) -> Vec<Arc<dyn Tester>> {
        self.testers.read().unwrap().values().cloned().collect()
    }

    pub fn list_modules(&self) -> Vec<ModuleInfo> {
        let mut infos = Vec::new();

        for c in self.observers.read().unwrap().values() {
            infos.push(ModuleInfo {
                id: c.id().to_string(),
                name: c.name().to_string(),
                version: c.version().to_string(),
                module_type: "observer".to_string(),
                source_system: c.source_system().to_string(),
                safety_classification: String::new(),
                environment_scope: String::new(),
            });
        }

        for t in self.testers.read().unwrap().values() {
            infos.push(ModuleInfo {
                id: t.id().to_string(),
                name: t.name().to_string(),
                version: t.version().to_string(),
                module_type: "tester".to_string(),
                source_system: t.source_system().to_string(),
                safety_classification: t.safety_class().to_string(),
                environment_scope: t.environment_scope().to_string(),
            });
        }

        infos.sort_by(|a, b| a.id.cmp(&b.id));
        infos
    }

    pub fn list_by_type(&self, module_type: &str) -> Vec<ModuleInfo> {
        self.list_modules()
            .into_iter()
            .filter(|m| m.module_type == module_type)
            .collect()
    }

    pub fn list_by_source_system(&self, system: &str) -> Vec<ModuleInfo> {
        self.list_modules()
            .into_iter()
            .filter(|m| m.source_system == system)
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testutil::{MockObserver, MockTester};

    fn make_registry() -> Registry {
        let reg = Registry::new();
        reg.register_observer(Arc::new(MockObserver::new("col.a")));
        reg.register_observer(Arc::new(MockObserver::new("col.b")));
        reg.register_tester(Arc::new(MockTester::safe("test.a")));
        reg
    }

    #[test]
    fn get_observer_found() {
        let reg = make_registry();
        let c = reg.get_observer("col.a").unwrap();
        assert_eq!(c.id(), "col.a");
    }

    #[test]
    fn get_observer_not_found() {
        let reg = make_registry();
        let err = reg.get_observer("nonexistent").err().expect("should fail");
        assert!(err.to_string().contains("not found"));
    }

    #[test]
    fn get_tester_found() {
        let reg = make_registry();
        let t = reg.get_tester("test.a").unwrap();
        assert_eq!(t.id(), "test.a");
    }

    #[test]
    fn get_tester_not_found() {
        let reg = make_registry();
        assert!(reg.get_tester("missing").is_err());
    }

    #[test]
    fn get_module_finds_observer() {
        let reg = make_registry();
        let m = reg.get_module("col.a").unwrap();
        assert_eq!(m.id(), "col.a");
    }

    #[test]
    fn get_module_finds_tester() {
        let reg = make_registry();
        let m = reg.get_module("test.a").unwrap();
        assert_eq!(m.id(), "test.a");
    }

    #[test]
    fn get_module_not_found() {
        let reg = make_registry();
        assert!(reg.get_module("unknown").is_err());
    }

    #[test]
    fn list_observers_returns_all() {
        let reg = make_registry();
        assert_eq!(reg.list_observers().len(), 2);
    }

    #[test]
    fn list_testers_returns_all() {
        let reg = make_registry();
        assert_eq!(reg.list_testers().len(), 1);
    }

    #[test]
    fn list_modules_sorted_by_id() {
        let reg = make_registry();
        let modules = reg.list_modules();
        assert_eq!(modules.len(), 3);
        // Verify sorted
        let ids: Vec<&str> = modules.iter().map(|m| m.id.as_str()).collect();
        let mut sorted = ids.clone();
        sorted.sort();
        assert_eq!(ids, sorted);
    }

    #[test]
    fn list_modules_has_correct_types() {
        let reg = make_registry();
        let modules = reg.list_modules();
        let col_count = modules
            .iter()
            .filter(|m| m.module_type == "observer")
            .count();
        let test_count = modules.iter().filter(|m| m.module_type == "tester").count();
        assert_eq!(col_count, 2);
        assert_eq!(test_count, 1);
    }

    #[test]
    fn list_by_type_observer() {
        let reg = make_registry();
        let cols = reg.list_by_type("observer");
        assert_eq!(cols.len(), 2);
        assert!(cols.iter().all(|m| m.module_type == "observer"));
    }

    #[test]
    fn list_by_type_tester() {
        let reg = make_registry();
        let testers = reg.list_by_type("tester");
        assert_eq!(testers.len(), 1);
    }

    #[test]
    fn list_by_source_system() {
        let reg = make_registry();
        // All mocks have source_system = "mock"
        let mock_modules = reg.list_by_source_system("mock");
        assert_eq!(mock_modules.len(), 3);

        let aws_modules = reg.list_by_source_system("aws");
        assert!(aws_modules.is_empty());
    }

    #[test]
    fn tester_info_includes_safety_and_scope() {
        let reg = make_registry();
        let modules = reg.list_modules();
        let tester_info = modules.iter().find(|m| m.module_type == "tester").unwrap();
        assert!(!tester_info.safety_classification.is_empty());
        assert!(!tester_info.environment_scope.is_empty());
    }
}
